File: 03_Stack/05_Daily_Temperatures.java
class Solution {
    public int[] dailyTemperatures(int[] temperatures) {
        int n = temperatures.length;

        Stack<int[]> st = new Stack<>();
        int[] ans = new int[n];

        for (int i = 0; i < n; i++) {
            int temp = temperatures[i];
            while (!st.isEmpty() && st.peek()[0] < temp) {
                int[] top = st.pop();
                ans[top[1]] = i - top[1];
            }
            st.push(new int[] { temp, i });
        }

        return ans;
    }
}

File: 03_Stack/05_Daily_Temperatures.cpp
class Solution
{
public:
    vector<int> dailyTemperatures(vector<int> &temperatures)
    {
        stack<pair<int, int>> st; // <temp, index>
        vector<int> ans(temperatures.size());

        for (int i = 0; i < temperatures.size(); i++)
        {
            int temp = temperatures[i];
            while (!st.empty() && st.top().first < temp)
            {
                ans[st.top().second] = i - st.top().second;
                st.pop();
            }
            st.push({temp, i});
        }

        return ans;
    }
};